### PR TITLE
✨ add storybook v7 support

### PR DESCRIPTION
**Pull request by nilshah98**
**Description -**
- [x] update `evalStorybookEnvironmentInfo` to grep storybook v7 version
- [x] update `evalStorybookStorySnapshots` to support fetching all stories even with `storyStorev7: true`
- [x] update packages to storybook v7 and update tests (`start-storybook` to `storybook dev`)
- [x] update minimum node version required to `16` and update github actions to use node `16` 
- [x] verify `cacheAllCSFFiles` with `storyStorev7: false` and also older storybook version
- [x] debug flaky tests

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -180,6 +183,9 @@ export function evalStorybookStorySnapshots({ waitFor }) {
   };
 
   return waitFor(async () => {
+    // uncache stories, if cached via storyStorev7: true
```
Comment (itsjwala) [inline src/utils.js:186]: does this work with `storyStorev7: false` ?

Comment (itsjwala) [reply src/utils.js:186]: Lets add a test for the same.

File: src/start.js [hunk 2]
```diff
@@ -32,11 +32,11 @@ export const start = command('start', {
   let { default: { spawn } } = yield import('cross-spawn');
   let { host, port } = flags;
 
-  let args = ['--ci', `--host=${host}`, `--port=${port}`, ...argv];
-  log.info(`Running "start-storybook ${args.join(' ')}"`);
+  let args = ['dev', '--ci', `--host=${host}`, `--port=${port}`, ...argv];
```
Comment (itsjwala) [inline src/start.js:35]: we'll need backward compatibility with v6.

**Review by itsjwala — COMMENTED** [review]
Looks good so far, nice!

* we'll need backward compatibility with v6 and have to figure out how we can add tests for same.
* we'll create E2E build on Percy and verify visual regression if any with v6 vs v7 and callout.
  * how many customers use node 14? we might need to have CI tests on node 14 as well.

File: package.json [hunk 3]
```diff
@@ -7,7 +7,7 @@
     "url": "https://github.com/percy/percy-storybook.git"
   },
   "engine": {
-    "node": ">=14"
+    "node": ">=16"
```
Comment (itsjwala) [inline package.json:10]: we don't need to have this compulsion, our code will still work with node 14.

(though does it let us run just verify) ?

Comment (nilshah98) [reply package.json:10]: Nope, storybook v7 is [**not** compatible with node 14](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#dropped-support-for-node-15-and-below).

Screenshot 2023-10-19 at 9 49 29 PM

File: src/start.js (hunk 2 above)
Comment (nilshah98) [reply src/start.js:35]: Yep, this is a little tricky.
I am unsure how to structure the SDK in that case.

File: src/utils.js (hunk 1 above)
Comment (nilshah98) [reply src/utils.js:186]: Will check this.
Also need to verify this with older versions of storybook

**Comment by nilshah98** [comment]
I have noticed that the tests are flaky.
I am not sure if this was the behaviour before as well, might need to debug this.
Example -
[Successful run](https://github.com/percy/percy-storybook/actions/runs/6577372147)
[Unsuccessful run](https://github.com/percy/percy-storybook/actions/runs/6577372147/job/17868869143?pr=829)

Both of them are for the same commit ids

File: src/utils.js (hunk 1 above)
Comment (nilshah98) [reply src/utils.js:186]: Refer - https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#v7-store-api-changes-for-addon-authors

File: test/storybook.test.js [hunk 4]
```diff
@@ -49,7 +55,11 @@ describe('percy storybook', () => {
   it('snapshots live urls', async () => {
     await storybook(['http://localhost:9000']);
 
-    expect(logger.stderr).toEqual([]);
+    // if there are stderr logs ensure it is only an acceptable warning
+    expect(logger.stderr).toEqual(logger.stderr.length ? [
+      '[percy] Waiting on a response from Storybook...'
```
Comment (itsjwala) [inline test/storybook.test.js:60]:
```suggestion
      '[percy] Waiting on a response from Storybook...'
```

why was this required? are we not waiting enough for storybook?

**Review by itsjwala — COMMENTED** [review]
looks good, some customary changes are pending.

apart from what we discussed in sync.

File: src/utils.js (hunk 1 above)
Comment (itsjwala) [reply src/utils.js:201]: v6 and v7 with `true` tests are covered, can you add tests for v7 with `false`

File: test/storybook.test.js (hunk 4 above)
Comment (nilshah98) [reply test/storybook.test.js:60]: We have added extra await for [checkStorybookVersion](https://github.com/percy/percy-storybook/blob/add-storybookv7-support/test/storybook.test.js#L20), this might be causing delays for storybook to start.
Hence, needed to add extra waiting check.

We already do something similar in [storybook-start.test.js](https://github.com/percy/percy-storybook/blob/master/test/storybook-start.test.js#L25)

File: src/utils.js (hunk 1 above)
Comment (nilshah98) [reply src/utils.js:201]: This will make the tests even more complicated.
We _might_ have to run all test cases again.

Why is this an issue, as this has been officially documented by storybook team, that we need to call `cacheAllCSFFiles` as documented above.

File: src/utils.js [hunk 5]
```diff
@@ -147,7 +162,10 @@ export async function* withPage(percy, url, callback, retry) {
 // Evaluate and return Storybook environment information from the about page
 /* istanbul ignore next: no instrumenting injected code */
 export function evalStorybookEnvironmentInfo({ waitForXPath }) {
-  return waitForXPath("//header[starts-with(text(), 'Storybook ')]", 5000)
+  let possibleEnvs = [];
```
[the remark this answers is not in the document]
Comment (nilshah98) [reply src/utils.js:165]: I think current tests are good enough.
We are already checking for `storybook/<number>`.

Current production is throwing `storybook/unknown` for storybook v7.

File: src/utils.js (hunk 1 above)
Comment (itsjwala) [reply src/utils.js:201]: we don't want a case where calling when its `false` and it errors out.

Though I'll keep this requirement optional if this is tested locally by creating a percy build.

File: src/utils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (itsjwala) [reply src/utils.js:165]: I meant check for equality of this xpath in the test case and if it returns correct version.
or another option is when posting the snapshot we intercept the network call to verify the environment details being sent.

I its already there we can skip, this is to avoid regression when storybook is updated locally and its go untested.

**Review by itsjwala — APPROVED** [review]
LGTM 🔥

File: src/utils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (nilshah98) [reply src/utils.js:165]: There is already test for this at -
https://github.com/percy/percy-storybook/blob/master/test/storybook.test.js#L245

[the remark this answers is not in the document]
Comment (itsjwala) [reply src/utils.js:165]: This is resolved in sync, there is already a test for this.
